### PR TITLE
feat(share/p2p/peers): Implement peer manager

### PR DESCRIPTION
**Pull request by walldiss**
## TODO
In future prs:

 - add backoff
 - add debug logs
 - add GC for old / unwanted pools
 - add metrics:
   - validation results rate[result_type, source:discovery/shrexsub]
   - peer retrival rate[shrexsub/discovery]
   - retrieval time hist
   - validation time hist
   - blacklisted peers count
   - blacklistedHash count

## Checklist

- [ ] New and updated code has appropriate documentation
- [ ] New and updated code has new and/or updated testing
- [ ] Required CI checks are passing
- [ ] Visual proof for any user facing features like CLI or documentation updates
- [ ] Linked issues closed with keywords

### CODE REVIEW

**Review by walldiss — COMMENTED** [review]
Comments from sync review

**Review by distractedm1nd — APPROVED** [review]

File: share/p2p/peers/manager.go [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+	connGater *conngater.BasicConnectionGater
+
+	// pools collecting peers from shrexSub
+	pools           map[string]*syncPool
+	poolSyncTimeout time.Duration
+	// fullNodes collects full nodes peer.ID found via discovery
+	fullNodes *pool
+
+	// peers that misbehaved
+	blacklistedPeers map[peer.ID]bool
```
Comment (vgonkivs) [inline share/p2p/peers/manager.go:66]: we can avoid this field by relying on ConnGater. Please check this [link](https://github.com/libp2p/go-libp2p/blob/635cc650cfb3bbbdb4bef6ae29f45e3c8966fb31/p2p/net/conngater/conngater.go#L277)

File: share/p2p/peers/pool.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+				continue
+			}
+			p.nextIdx = len(newList)
+		}
+	}
+	p.peersList = newList
+}
+
+// checkHasPeers will check and indicate if there are peers in the pool.
+func (p *pool) checkHasPeers() {
```
Comment (vgonkivs) [inline share/p2p/peers/pool.go:145]: nit: would name it `hasPeers()`

**Review by renaynay — APPROVED** [review]

File: share/p2p/peers/pool.go (hunk 2 above)
Comment (renaynay) [reply share/p2p/peers/pool.go:145]: confusing bc hasPeer is field

File: share/p2p/peers/manager.go (hunk 1 above)
Comment (renaynay) [reply share/p2p/peers/manager.go]: done

**Review by vgonkivs — APPROVED** [review]

**Review by renaynay — APPROVED** [review]

**Review by distractedm1nd — APPROVED** [review]

**Pull request merged**